### PR TITLE
Deprecate config item `max-memory` and add items `server-memory-quota` and `memory-usage-alarm-ratio`

### PR DESCRIPTION
**Pull request by CharLotteiu**
### What is changed, added, or deleted? (Required)

Deprecate config item max-memory and add items server-memory-quota and memory-usage-alarm-ratio (only in dev)

max-memory has been deprecated and use server-memory-quota instead now. (only in dev)
memory-usage-alarm-ratio is introduced in this pr pingcap/tidb#18858.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/4977
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [x] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by wshwsh12 — COMMENTED** [review]
LGTM

File: tidb-configuration-file.md [hunk 1]
```diff
@@ -286,11 +286,22 @@ Configuration items related to performance.
...
+### `server-memory-quota`
 
-- The maximum memory limit for the Prepared Least Recently Used (LRU) caching. If this value exceeds `performance.max-memory * (1 - prepared-plan-cache.memory-guard-ratio)`, the elements in the LRU are removed.
-- Default value: `0`
-- This configuration takes effect when `prepared-plan-cache.enabled` is `true` (default). When the size of the LRU is greater than `prepared-plan-cache.capacity`, the elements in the LRU are also removed.
+> **Warning:**
+>
+> `server-memory-quota` is still an experimental feature. It is **NOT** recommended that you use it in a production environment.
+
++ The memory usage limit of tidb-server instances. <!-- As of TiDB v5.0 --> This configuration item completely replaces the original [`max-memory`](https://docs.pingcap.com/zh/tidb/stable/tidb-configuration-file#max-memory).
```
Comment (TomShawn) [inline tidb-configuration-file.md:295]:
```suggestion
+ The memory usage limit of tidb-server instances. <!-- New in TiDB v5.0 --> This configuration item completely supercedes the previous [`max-memory`](https://docs.pingcap.com/tidb/stable/tidb-configuration-file#max-memory).
```

File: tidb-configuration-file.md [hunk 2]
```diff
@@ -286,11 +286,22 @@ Configuration items related to performance.
...
++ Default value: `0` (in bytes)
```
Comment (TomShawn) [inline tidb-configuration-file.md:296]:
```suggestion
+ Default value: `0` (in bytes), which means no memory limit.
```

File: tidb-configuration-file.md [hunk 3]
```diff
@@ -286,11 +286,22 @@ Configuration items related to performance.
...
+
+### `memory-usage-alarm-ratio`
+
++ TiDB triggers an alarm when the memory usage of tidb-server instance exceeds a certain threshold. The valid value for this configuration item ranges from `0` to `1`. If it is configured as `0` or `1`, then the memory threshold alarm feature is disabled.
```
Comment (TomShawn) [inline tidb-configuration-file.md:300]:
```suggestion
+ TiDB triggers an alarm when the memory usage of tidb-server instance exceeds a certain threshold. The valid value for this configuration item ranges from `0` to `1`. If it is configured as `0` or `1`, this alarm feature is disabled.
```

File: tidb-configuration-file.md [hunk 4]
```diff
@@ -286,11 +286,22 @@ Configuration items related to performance.
...
++ Default value: `0.8`
++ When the memory usage alarm is enabled, if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is not set, then the threshold of memory usage equals the value of `memory-usage-alarm-ratio` times the system memory size; if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is set to greater than 0, then the threshold of memory usage equals the value of `memory-usage-alarm-ratio` times the value of `server-memory-quota`.
```
Comment (TomShawn) [inline tidb-configuration-file.md:302]: When `time` means `乘以`, it is used as a verb.
```suggestion
+ When the memory usage alarm is enabled, if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is not set, then the threshold of memory usage is ``` the `memory-usage-alarm-ratio` value * the system memory size ```; if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is set to a value greater than 0, then the threshold of memory usage is ``` the `memory-usage-alarm-ratio` value * the `server-memory-quota` value ```.
```

File: tidb-configuration-file.md [hunk 5]
```diff
@@ -286,11 +286,22 @@ Configuration items related to performance.
...
++ When TiDB detects that the memory usage of the tidb-server exceeds the threshold, it considers that there might be a risk of OOM. Therefore, it records 10 SQL statements with the highest memory usage, 10 SQL statements with the longest running time, and the heap profile among all SQL statements currently being executed to the directory [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) and outputs a log containing the keyword `tidb-server has the risk of OOM`.
```
Comment (TomShawn) [inline tidb-configuration-file.md:303]:
```suggestion
+ When TiDB detects that the memory usage of the tidb-server instance exceeds the threshold, it considers that there might be a risk of OOM. Therefore, it records ten SQL statements with the highest memory usage, ten SQL statements with the longest running time, and the heap profile among all SQL statements currently being executed to the directory [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) and outputs a log containing the keyword `tidb-server has the risk of OOM`.
```

**Review by TomShawn — COMMENTED** [review]
Some minor comments on wording~

File: tidb-configuration-file.md [hunk 6]
```diff
@@ -286,11 +286,22 @@ Configuration items related to performance.
...
+> `server-memory-quota` is still an experimental feature. It is **NOT** recommended that you use it in a production environment.
+
++ The memory usage limit of tidb-server instances. <!-- New in TiDB v5.0 --> This configuration item completely supersedes the previous [`max-memory`](https://docs.pingcap.com/tidb/stable/tidb-configuration-file#max-memory).
++ Default value: `0` (in bytes), which means no memory limit.
+
+### `memory-usage-alarm-ratio`
+
++ TiDB triggers an alarm when the memory usage of tidb-server instance exceeds a certain threshold. The valid value for this configuration item ranges from `0` to `1`. If it is configured as `0` or `1`, this alarm feature is disabled.
++ Default value: `0.8`
++ When the memory usage alarm is enabled, if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is not set, then the threshold of memory usage is ``` the `memory-usage-alarm-ratio` value * the system memory size ```; if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is set to a value greater than 0, then the threshold of memory usage is ``` the `memory-usage-alarm-ratio` value * the `server-memory-quota` value ```.
```
Comment (TomShawn) [inline tidb-configuration-file.md:302]:
```suggestion
+ When the memory usage alarm is enabled, if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is not set, then the threshold of memory usage is ```the `memory-usage-alarm-ratio` value * the system memory size```; if [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) is set to a value greater than 0, then the threshold of memory usage is ```the `memory-usage-alarm-ratio` value * the `server-memory-quota` value```.
```

**Review by TomShawn — APPROVED** [review]
LGTM

**Comment by TomShawn** [comment]
Follow-up: remove `<!-- New in 5.0 -->`

**Pull request merged**